nat-lab/tests/test_pq.py

```diff
@@ -130,3 +131,111 @@ async def test_pq_vpn_connection(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    stun_limits=ConnectionLimits(1, 1),
+                    nlx_1_limits=ConnectionLimits(2, 2),
+                ),
+                is_meshnet=False,
+            ),
+            "10.0.254.1",
+            marks=pytest.mark.linux_native,
+        ),
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.WINDOWS_VM_1,
+                adapter_type=AdapterType.WindowsNativeWg,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.WINDOWS_VM_1,
+                    stun_limits=ConnectionLimits(1, 1),
+                    nlx_1_limits=ConnectionLimits(2, 2),
+                ),
+                is_meshnet=False,
+            ),
+            "10.0.254.7",
+            marks=[
+                pytest.mark.windows,
+            ],
+        ),
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.WINDOWS_VM_1,
+                adapter_type=AdapterType.WireguardGo,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.WINDOWS_VM_1,
+                    stun_limits=ConnectionLimits(1, 1),
+                    nlx_1_limits=ConnectionLimits(2, 2),
+                ),
+                is_meshnet=False,
+            ),
+            "10.0.254.7",
+            marks=[
+                pytest.mark.windows,
+            ],
+        ),
+        # pytest.param(
+        #     SetupParameters(
+        #         connection_tag=ConnectionTag.MAC_VM,
+        #         adapter_type=AdapterType.BoringTun,
+        #         connection_tracker_config=generate_connection_tracker_config(
+        #             ConnectionTag.MAC_VM,
+        #             stun_limits=ConnectionLimits(1, 1),
+        #             nlx_1_limits=ConnectionLimits(2, 2),
+        #         ),
+        #         is_meshnet=False,
+        #     ),
+        #     "10.0.254.7",
+        #     marks=pytest.mark.mac,
+        # ),
+    ],
+)
+async def test_pq_vpn_rekey(
+    alpha_setup_params: SetupParameters,
+    public_ip: str,
+) -> None:
+    # Set rekey interval to some small value
+    alpha_setup_params.features.post_quantum_vpn.rekey_interval_s = 4
+
+    async with AsyncExitStack() as exit_stack:
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, [alpha_setup_params])
+        )
+
+        client_conn, *_ = [conn.connection for conn in env.connections]
+        client_alpha, *_ = env.clients
+
+        ip = await testing.wait_long(stun.get(client_conn, config.STUN_SERVER))
+        assert ip == public_ip, f"wrong public IP before connecting to VPN {ip}"
+
+        await _connect_vpn_pq(
+            client_conn,
+            client_alpha,
+        )
+
+        # Sleep for 6s to allow for the rekey to happen
+        await asyncio.sleep(6)
```

Comment:
isn't this racy condition, what is making sure that rekey happened in 6 seconds? maybe so event happens during that time?

Reply:
I'm setting the REKEY interval to 4s here https://github.com/NordSecurity/libtelio/pull/518/files/6ee7026e3ea09d2d3e1091dd6d71ee24931680ab#diff-d71adf294c60fb47c1fdaabf09dfb47c83a4566748a530105ca35e310f7f49fdR219

Reply:
There's no other way, the client is responsible for periodic rekey so whatever we feed the client it's fine as long as the VPN connection is not broken.
I believe that without rekeys at all the server has a hardcoded timeout on the connection but it is like 5 minutes or something

Reply:
Oh, do you mean something like inspecting the logs instead of sleeping? Good point actually

Reply:
Also, I think, you can easily ping PHOTO album even without vpn connection. So maybe do `stun.get` to check wether we still have VPN server IP as our public ip

Reply:
I've added the wait for log and it was not appearing. Good catch